docs/Developers/NextGen_Core.rst

```diff
@@ -0,0 +1,303 @@
+    # more analysis after receving its results
+    class SomeOtherBear(Bear):
+        BEAR_DEPS = {SomeBear}
+
+        def analyze(self, *args):
+            return "More analysis"
+
+        def generate_tasks(self):
+            result_length = sum(len(results)
+                                for results in self.dependency_results.values())
+            return (((i,), {}) for i in range(result_length))
+
+Running the Core
+----------------
+
+The `run(bears, result_callback, cache=None, executor=None)` method in
+``Core.py`` is used to set up a coala session and run it.
+- ``bears``: contains all the instances of bears to run.
+- ``result_callback``: a function that is called only when the results from
+the bears are available.
+- ``cache``: a cache used by bears to speed up runs.
+- ``executor``: a custom executor used to run the bears. By default
+``ProcessPoolExecutor`` is used.
+
+The new core comes with the following new additions:
+
+- **Dependency Resolution:** The ``initialize_dependencies`` method in
+  ``Core.py`` receives the bears that are to be run and processes bear
+  dependencies using a consumer-based system so that each dependency bear has
+  only one instance per section and file-dict. It returns a set of dependency
+  bears along with those bears that don't have any dependencies or whose
+  dependencies have been resolved (these are the ones that are scheduled to be
+  run). The bears are resolved by traversing each edge of the dependency graph
+  and instantiating dependency bears accordingly one per section. Before the
+  bears are run we initialize the dependency tracking in the ``__init__`` method
+  of the class ``Session`` which is responsible for running coala sessions. The
+  ``dependency_tracker`` object is used while scheduling and cleaning up bears
+  and also while finishing bear tasks.
+- **Task Scheduling:** The bears that have no dependencies or whose
+  dependencies have been resolved, only their tasks will be scheduled for
+  execution. Before executing any task coala looks it up in the cache.
+  In case of a hit, the existing results that are stored in the cache for the
+  corresponding task arguments are called using ``execute_task_with_cache``
+  method. In case of a miss or if coala is run without a cache the task is
+  executed. The bears without any running tasks are cleaned up from the state of
+  an ongoing run by resolving its dependencies, scheduling dependent bears
+  and removing the bear from the ``running_tasks`` dict. When a task belonging
+  to a particular bear completes, then ``_finish_task`` method is responsible
+  for checking whether the bear execution is complete and the result generated
+  from the task have been returned. If any new tasks are to be scheduled then
+  this method is responsible for doing so after the dependencies have been
+  resolved.
+- **result_callback :** Each of the bears scheduled to run for a session has
```

Comment:
Again I think that even this one can be removed by moving any new info that is present here to the **Easier Interface** paragraph where you explain `result_callback` as one of the parameters that is passed to `run()`.

Reply:
Its already mentioned in Easier Interface section. Since that section is about new interface feature I don't think implementation details like these should be discussed in details over there its better to have a separate section for the details.